### ISSUE TITLE
RestClient support customiser

### ISSUE THREAD

**Issue by M-Whitaker** [spring-projects/spring-hateoas#2038]
As of Spring Boot 3.2.0 RestClient has been added via Spring Framework to replace WebClient. Is there a replacement for hypermediaWebClientCustomizer?

**Comment by odrotbohm** [comment]
`RestClient` works with a `RestTemplate` under the covers, so all hypermedia-related customizations would have to be (and are, in fact) applied to that, right?

**Comment by M-Whitaker** [comment]
Yes this is my understanding. Currently I have a bean like this:


```
@Bean
  public RestClientCustomizer hypermediaRestClientCustomizer(WebConverters webConverters) {
    return builder -> builder.messageConverters(webConverters::augmentClient);

```

Which works fine, as you said. Is there plans to put a wrapper around this like has been done for resttemplate & web client?

**Comment by bertsanders** [comment]
This worked for me, but only if I use 'RestClient.ResponseSpec.toEntity()'.  `RestClient.ResponseSpec.body()` failed to parse the response to a PagedModelType.

**Comment by pcornelissen** [comment]
Maybe there should be at least a section in the documentation what to do or expect when you want to  use hateoas with RestClient.
I don't think that many/all users of the hateoas Lib with "regular" spring knowledge know what to do in this case.

**Comment by snicoll** [comment]
Isn't this related to https://github.com/spring-projects/spring-boot/issues/49806 as well @odrotbohm

**Comment by maddenj-ie** [comment]
Using Spring Boot 4.1.1, Spring Hateous 3.1.2 (managed via boot)

```
@Bean
    RestClient myRestClient(
            final WebConverters webConverters,
            final RestClient.Builder builder) {
        return builder
            .configureMessageConverters(
                converters -> converters.configureMessageConvertersList(webConverters::augmentClient))
            .build();
    }
```

Is this the appropriate way to apply the hypermedia related customizations?

**Comment by pcornelissen** [comment]
Yes, that works and I think that is the intended way by the spring-gods. I checked it against Boot 4.1.1 and Spring HATEOAS 3.1.2 with a small MVC app serving HAL and a `RestClient` calling it, and it does the right thing.

Why: Boot's `HttpMessageConvertersRestClientCustomizer` already calls `restClientBuilder.configureMessageConverters(...)`, and `DefaultClientHttpMessageConvertersCustomizer` calls `builder.registerDefaults()` inside it. Your configurer gets chained onto that one via `Consumer.andThen`, and `configureMessageConvertersList(...)` runs in `DefaultHttpMessageConverters.Builder#build()` against the finished converter list. That list holds the `JacksonJsonHttpMessageConverter` Boot created through `withJsonConverter(...)`, which is exactly what `WebConverters.augmentClient(...)` goes looking for, so the HAL mappers get registered on it in place.

I got `EntityModel<T>`, `CollectionModel<EntityModel<T>>` and `PagedModel<EntityModel<T>>` back correctly, with `.body(...)` and `.toEntity(...)` alike, so the `body()` problem @bertsanders ran into in 2023 seems to be gone. Plain non-hypermedia DTOs on the same client keep working too. Drop the customization and the same calls come back with an empty `Links` collection and no error, which is the regression from spring-projects/spring-boot#49806.

**Three caveats, though.**

The first one is that this only holds for the Boot-managed `RestClient.Builder`. `DefaultRestClientBuilder#initMessageConverters()` skips `registerDefaults()` the moment any converter configurer is present, and with the injected builder it's Boot's own customizer that supplies the defaults. Move the same snippet to a standalone `RestClient.builder()` and it blows up:

```
UnknownContentTypeException: Could not extract response: no suitable HttpMessageConverter found
for response type [org.springframework.hateoas.EntityModel<...>] and content type [application/hal+json]
```

There you have to register the defaults yourself:

```java
RestClient.builder()
    .configureMessageConverters(c -> c.registerDefaults()
        .configureMessageConvertersList(webConverters::augmentClient))
    .build();
```

Second, if you need this for more than one client, a `RestClientCustomizer` bean is tidier. Same effect, applied to every auto-configured builder, and since that builder is `prototype`-scoped there's no shared state to worry about:

```java
@Bean
RestClientCustomizer hypermediaRestClientCustomizer(WebConverters webConverters) {
    return builder -> builder.configureMessageConverters(
        converters -> converters.configureMessageConvertersList(webConverters::augmentClient));
}
```

Third, and this one might deserve its own issue: `WebConverters#augment` does

```java
var converter = converters.stream()
        .filter(it -> JacksonJsonHttpMessageConverter.class.equals(it.getClass()))
        ...
        .orElseGet(() -> new JacksonJsonHttpMessageConverter(mapper));
```

The converter from `orElseGet` gets configured and then dropped on the floor, it never makes it into `converters`. Combine that with the exact class match in the filter and you get a silent no-op whenever the JSON converter in the list happens to be a subclass of `JacksonJsonHttpMessageConverter`. That's not hypothetical: someone passing their own converter to `ClientBuilder#withJsonConverter`, or `TypeConstrainedJacksonJsonHttpMessageConverter`, or Spring Data REST's `AlpsJacksonJsonHttpMessageConverter` all land there. I reproduced it with a two-line subclass. No exception, nothing in the log, just models without links. Adding the fallback converter to the list, or loosening the filter to `instanceof`, would at least make the failure visible.

None of that changes the answer to your question, but a `HypermediaRestClientConfigurer` next to `HypermediaRestTemplateConfigurer` and `HypermediaWebClientConfigurer`, plus a paragraph in the docs, would spare the next person from having to find `WebConverters` on their own. Which is more or less what this issue has been asking for since 2023.
